### ISSUE TITLE
RuntimeError: memory access out of bounds

### ISSUE THREAD

**Issue by acosta-leandro** [martinjrobins/diffeq-js#3]
I'm trying to implement this equation using the new version '^0.1.5-alpha.0' and encountering the same problem on [Diffeq-web](https://martinjrobins.github.io/diffeq-web/).

Could you please take a look and check it?

Eq:
`in = [p1, p2, p3, p4, p5, p6, p7, p8, p9, voltageClampELeak,
  voltageClampELeakEst, voltageClampRSealMOhm, voltageClampRSealEstMOhm,
  voltageClampCPrs, voltageClampCPrsEst, voltageClampRSeriesMOhm,
  voltageClampRSeriesEstMOhm, voltageClampCmEst, voltageClampAlphaPPercentage,
  voltageClampAlphaRPercentage, voltageClampTauClamp, voltageClampTauOut,
  voltageClampTauRs, voltageClampTauSum, voltageClampVOffsetEff, cellCm,
  extraKo, kiKi ]
  idealVC {
 -80
- 40 * heaviside (t - 250)
+ 40 / 400 * (t - 255) * heaviside(t - 255) - 40 / 400 * (t - 655) * heaviside(t - 655) 
+ 120 * heaviside (t - 855)
- 160 * heaviside (t - 1855)
+ 40 * heaviside (t - 2355)
+ 40 * heaviside (t - 3355)
- 20 * heaviside (t - 3855)
+ 40 * heaviside (t - 4355)
- 20 * heaviside (t - 4855)
+ 40 * heaviside (t - 5355)
- 20 * heaviside (t - 5855)
+ 40 * heaviside (t - 6355)
- 20 * heaviside (t - 6855)
+ 40 * heaviside (t - 7355)
- 20 * heaviside (t - 7855)
+ 20 * heaviside (t - 8355)
- 40 * heaviside (t - 8855)
+ 20 * heaviside (t - 9355)
- 40 * heaviside (t - 9855)
+ 20 * heaviside (t - 10355)
- 40 * heaviside (t - 10855)
+ 20 * heaviside (t - 11355)
- 40 * heaviside (t - 11855)
+ 20 * heaviside (t - 12355)
- 40 * heaviside (t - 12855)
+ 120 * heaviside (t - 13355)
- 110 * heaviside (t - 13855)
- 50 / 100 * (t - 13865) * heaviside(t - 13865) + 50 / 100 * (t - 13965) * heaviside(t - 13965) 
+ 40 * heaviside (t - 14355)
}
  p1 { 2.07e-3 }
  p2 { 7.17e-2 }
  p3 { 3.44e-5 }
  p4 { 6.18e-2 }
  p5 { 4.18e-1 }
  p6 { 2.58e-2 }
  p7 { 4.75e-2 }
  p8 { 2.51e-2 }
  p9 { 50 }
  voltageClampELeak { -80 }
  voltageClampELeakEst { -80 }
  voltageClampRSealMOhm { 5000 }
  voltageClampRSealEstMOhm { 5000 }
  voltageClampCPrs { 4 }
  voltageClampCPrsEst { 4 }
  voltageClampRSeriesMOhm { 30 }
  voltageClampRSeriesEstMOhm { 30 }
  voltageClampCmEst { 20 }
  voltageClampAlphaPPercentage { 70 }
  voltageClampAlphaRPercentage { 70 }
  voltageClampTauClamp { 0.001 }
  voltageClampTauOut { 0.0075 }
  voltageClampTauRs { 0.1 }
  voltageClampTauSum { 0.001 }
  voltageClampVOffsetEff { 0 }
  voltageClampRSeries { voltageClampRSeriesMOhm * 0.001 }
  voltageClampRSeriesEst { voltageClampRSeriesEstMOhm * 0.001 }
  voltageClampAlphaP { voltageClampAlphaPPercentage / 100 }
  voltageClampAlphaR { voltageClampAlphaRPercentage / 100 }
  voltageClampGLeak { 1 / (voltageClampRSealMOhm * 0.001) }
  voltageClampGLeakEst { heaviside(voltageClampRSealEstMOhm - 1e-6) * (1 / (voltageClampRSealEstMOhm * 0.001)) }
  cellCm { 20 }
  extraKo { 5.4 }
  extraNao { 140 }
  kiKi { 130 }
  naiNai { 10 }
  physR { 8.314472 }
  physT { 310 }
  physF { 9.64853415000000041e1 }
  physRTF { physR * physT / physF }
  erevEK { physRTF * log(extraKo / kiKi) }
  erevENa { physRTF * log(extraNao / naiNai) }
  u_i {
      membraneV = -80,
      ikrI = 0,
      ikrCI = 0,
      ikrO = 0,
      ikrIi = 0,
      ikrCIi = 0,
      ikrOi = 0,
      voltageClampVClamp = -80,
      voltageClampVP = -80,
      voltageClampVEst = -80,
      voltageClampIOut = 0,
      voltageClampIInP = 0
  }
  k12 { p1 * exp(p2 * membraneV) }
  k12i { p1 * exp(p2 * idealVC) }
  k14 { p7 * exp(-p8 * membraneV) }
  k14i { p7 * exp(-p8 * idealVC) }
  k21 { p3 * exp(-p4 * membraneV) }
  k21i { p3 * exp(-p4 * idealVC) }
  k41 { p5 * exp(p6 * membraneV) }
  k41i { p5 * exp(p6 * idealVC) }
  ikrC { 1 - ikrCI - ikrI - ikrO }
  ikrCi { 1 - ikrCIi - ikrIi - ikrOi }
  ikrIKr { p9 * ikrO * (membraneV - erevEK) }
  ikrIKrI { p9 * ikrOi * (idealVC - erevEK) }
  membraneIIdeal { ikrIKrI }
  membraneIIon { ikrIKr }
  voltageClampILeak { voltageClampGLeak * (membraneV - voltageClampELeak) }
  dudt_i {
      dmembraneVdt = -80,
      dikrIdt = 0,
      dikrCIdt = 0,
      dikrOdt = 0,
      dikrIidt = 0,
      dikrCIidt = 0,
      dikrOidt = 0,
      dvoltageClampVClampdt = -80,
      dvoltageClampVPdt = -80,
      dvoltageClampVEstdt = -80,
      dvoltageClampIOutdt = 0,
      dvoltageClampIInPdt = 0
  }
  F_i {
      dmembraneVdt,
      dikrIdt,
      dikrCIdt,
      dikrOdt,
      dikrIidt,
      dikrCIidt,
      dikrOidt,
      dvoltageClampVClampdt,
      dvoltageClampVPdt,
      dvoltageClampVEstdt,
      dvoltageClampIOutdt,
      dvoltageClampIInPdt
  }
  voltageClampIIn { ((voltageClampVP - membraneV + voltageClampVOffsetEff) / voltageClampRSeries + voltageClampCPrs * dvoltageClampVPdt - voltageClampCPrsEst * dvoltageClampVClampdt - voltageClampCmEst * dvoltageClampVEstdt) / cellCm }
  G_i {
      (voltageClampVP + voltageClampVOffsetEff - membraneV) / (cellCm * voltageClampRSeries) - (membraneIIon + voltageClampILeak) / cellCm,
      -k14 * ikrI + k41 * ikrO + k12 * ikrCI - k21 * ikrI,
      -k12 * ikrCI + k21 * ikrI + k41 * ikrC - k14 * ikrCI,
      -k21 * ikrO + k12 * ikrC + k14 * ikrI - k41 * ikrO,
      -k14i * ikrIi + k41i * ikrOi + k12i * ikrCIi - k21i * ikrIi,
      -k12i * ikrCIi + k21i * ikrIi + k41i * ikrCi - k14i * ikrCIi,
      -k21i * ikrOi + k12i * ikrCi + k14i * ikrIi - k41i * ikrOi,
      (idealVC + (voltageClampIInP * cellCm * voltageClampAlphaR + voltageClampCmEst * dvoltageClampVEstdt * voltageClampAlphaP) * voltageClampRSeriesEst - voltageClampVClamp) / voltageClampTauSum,
      (voltageClampVClamp - voltageClampVP) / voltageClampTauClamp,
      heaviside(voltageClampCmEst - 1e-6) *  heaviside(voltageClampRSeriesEst - 1e-6) * ((idealVC - voltageClampVEst) / ((1 - voltageClampAlphaP) * voltageClampCmEst * voltageClampRSeriesEst)),
      (voltageClampIIn - voltageClampIOut) / voltageClampTauOut,
      (voltageClampIIn - voltageClampIInP) / voltageClampTauRs
  }
  voltageClampIOutPA { voltageClampIOut * cellCm }
  voltageClampIPost { voltageClampIOut - voltageClampGLeakEst * (idealVC - voltageClampELeakEst) / cellCm }
  voltageClampIPostPA { voltageClampIPost * cellCm }
  out_i {
      voltageClampIOut,
      idealVC,
      voltageClampIIn,
      voltageClampIPostPA,
      voltageClampIPost,
      membraneV,
      membraneIIdeal,
      membraneIIon,
  }
  `
  
![image](https://github.com/user-attachments/assets/182b2c7b-01d8-41bc-aea3-6387d57cc68a)

**Comment by martinjrobins** [comment]
hi @acosta-leandro, your model is fine, the issue is in diffeq-js, for some reason calling `solver.destroy()` is crashing the app when you swap between the default model to your particular model. I'm not sure the ultimate cause of this, but for the momoment I have commented out this line so that its usable for you

**Comment by acosta-leandro** [comment]
Could you provide an update on the npm package? I'm using it in my app.

Thank you

**Comment by martinjrobins** [comment]
The workaround mentioned above is just for the web app, not the npm package

**Comment by martinjrobins** [comment]
is the issue occuring in your app? In my web app, the issue occurs when you recompile a new model, then call solver.destroy() on a solver that was defined using the previous model

**Comment by acosta-leandro** [comment]
Yes, it's happening using the npm package as well. I do not invoke solver.destroy(), so perhaps it has been invoked by the package itself.

**Comment by martinjrobins** [comment]
This is probably a different bug then. Can you give me a minimal reproduction of the bug, or any other information?

**Comment by acosta-leandro** [comment]
I just updated it [here](https://github.com/acosta-leandro/test-diffeq-js). After clicking the button, it will compile and then solve. During the solving process, it has the error.

**Comment by acosta-leandro** [comment]
@martinjrobins Is there any update on this issue?

**Comment by martinjrobins** [comment]
sorry for the delay on this one, I've been on leave for most of the last 3 weeks and am at a conference this week. I'll look at it as soon as I can

**Comment by martinjrobins** [comment]
ok, I've had a chance to look at this. The issue is that you are putting a very large max time value into the solve, and basically you are running out of memory as the solver is trying to return a very large solution vector. Do you need to have `maxTime` that large (14455)?

**Comment by acosta-leandro** [comment]
Yes, this large max time was determined based on one of our experiments.

During my analysis of the browser's memory heap, I observed that the previous Float64Array instances are not being garbage collected after executing a new solve. This leads to a failure in the application just before Snapshot 3.

Here's a summary of my snapshots:

Snapshot 1: 14 Float64Array instances, totaling 57.9 MB.
Snapshot 2: Increased to 56 Float64Array instances, totaling 62.9 MB.
Snapshot 3: Further increased to 122 Float64Array instances, totaling 69.3 MB.

Test: Same equation, but using a smaller voltage protocol (3 steps) and a max time of 1500. Compile once, change parameters, and it crashes after several solve calls.

<img width="659" alt="1" src="https://github.com/user-attachments/assets/4aa3cad4-ca3b-4498-a7d0-26f9fb300d71">
<img width="659" alt="2" src="https://github.com/user-attachments/assets/b00f5505-9fd0-4d8d-827b-169485ddd0a3">
<img width="659" alt="3" src="https://github.com/user-attachments/assets/160edb15-a6a0-4442-b16f-762ca3f8c9b5">

**Comment by martinjrobins** [comment]
all of the diffsl classes have a `destroy()` method that will clean up their memory (web assembly doesn't have a GC). As noted earlier, the `solver.destroy()` method seems to have some issues though. But you might only need the `Vector::destroy` method to clean up. However, if a single solve takes 57.9 MB, and your required max-time is 10 times this that means 0.5 GB of memory to hold your solution, do you need this many points? You can pass a fixed number of timepoints to the solver to reduce the amount of points to hold your solution.

**Comment by kwabenantim** [comment]
The behaviour seems to depend on the order of inputs.

Using auto timesteps with the original model above, the `memory access out of bounds` errors appear to go away when the inputs are put in the same order as they are defined in the model.

Using fixed timesteps, it throws a new error:
```
[IDAS ERROR]  IDASolve 
  At t = 1.72449e-06, mxstep steps taken before reaching tout
SUNDIALS_ERROR: IDASolve() failed with retval = -1
```
This error also goes away when the inputs are ordered as they appear in the model.

I have tried to reduce the model to the smaller version below which still throws a`SUNDIALS_ERROR` when the input order is different (although it doesn't have the memory issues of the original model). 

Using fixed timesteps, an input order `in = [ P7, P5, P6 ]`, and input values `[0.0251, 0.418, 0.0258]`, it throws a `SUNDIALS_ERROR `. 

On the other hand, it runs fine when the inputs are ordered as `in = [ P5, P6, P7 ]` with the same values `[0.418, 0.0258, 0.0251]`.

Interestingly, both input orders work fine if the order in the model itself is changed to `P7 {...} P5 {...} P6 {...}`.

```js
import {
  compileModel,
  Options,
  Solver,
  Vector,
} from "@martinjrobins/diffeq-js";

var model = `
in = [ ]

VC { 40 }  

P2 { 0.0717 }
P7 { 0.0251 }
P5 { 0.418 }
P6 { 0.0258 }

Cm { 20 }
ELeak { -80.0 }
TauClamp { 0.001 }
TauRs { 0.1 }
TauSum { 0.001 }
RSeries { 0.03 }
RSeriesEst { 0.03 }
GLeak { 0.2 }

u_i {
  V = -80,
  O = 0,
  VClamp = -80, 
  VP = -80, 
  IOut = 0, 
  IInP = 0, 
}

dudt_i {
  diffV = 0,
  diffO = 0,
  diffVClamp = 0,
  diffVP = 0,
  diffIOut = 0,
  diffIInP = 0,
}

DotVP { (VClamp - VP) / TauClamp } 
DotVClamp { (VC + (IInP * Cm) * RSeriesEst - VClamp) / TauSum } 
IIn { ((VP - V) / RSeries + DotVP - DotVClamp) / Cm } 
ILeak { GLeak * (V - ELeak) } 

K12 { exp(P2 * V) } 
K41 { P5 * exp(P6 * V) } 

F_i {
  diffV,
  diffO,
  diffVClamp,
  diffVP,
  diffIOut,
  diffIInP,
}

G_i {
  (VP - V) / (Cm * RSeries) - ILeak / Cm,
  -O + K12 - K41 * O,
  DotVClamp,
  DotVP,
  (IIn - IOut),
  (IIn - IInP) / TauRs,
}

out_i { V }
`;

export const run = async (ordered) => {
  const params = {
    ordered: {
      P5: 0.418,
      P6: 0.0258,
      P7: 0.0251,
    },
    unordered: {
      P7: 0.0251,
      P5: 0.418,
      P6: 0.0258,
    },
  };

  const key = ordered ? "ordered" : "unordered";
  console.log(params[key]);

  // Model
  console.log("Model");
  const param_names = Object.keys(params[key]);
  model = model.replace("in = [ ]", `in = [ ${param_names.join(", ")} ]`);
  console.log(param_names);
  console.log(model);

  // Compile
  try {
    console.log("Compile");
    await compileModel(model);
  } catch (e) {
    console.log(e);
  }

  // Inputs
  console.log("Inputs:");
  const param_values = Object.values(params[key]);
  const inputs = new Vector(param_values);
  console.log(param_values);

  // Times
  console.log("Times:");
  const max_time = 1500;
  const steps = 1000;
  const t_arr = Array.from({ length: steps }, (_, i) => (max_time / steps) * i);
  const times = new Vector(t_arr);
  console.log("Steps: ", t_arr.length);
  console.log(t_arr.at(-1));
  console.log(t_arr);

  // Solve
  const outputs = new Vector([]);
  const options = new Options({ fixed_times: true });
  const solver = new Solver(options);

  try {
    console.log("Solve");
    solver.solve(times, inputs, outputs);

    const o_arr = outputs.getFloat64Array();
    console.log(o_arr);
  } catch (e) {
    console.log(e);
  }
};
```

**Comment by martinjrobins** [comment]
Thanks @kwabenantim, I'll have a look at this. Tbh I don't think I've ever tried the case where in the inputs are out of order in terms of how they are defined in the model, so it is very likely I've done something stupid here!

**Comment by martinjrobins** [comment]
I was indeed making the assumption that the ordering was consistent.... I'll put in a fix

**Comment by martinjrobins** [comment]
ok, the backend server has been updated with the fix, so cross fingers this should solve your problem

**Comment by kwabenantim** [comment]
Thanks @martinjrobins! The out of order inputs are working fine now. Is it working okay on the web app @acosta-leandro ?

**Comment by acosta-leandro** [comment]
It’s already solved. Thank you
